### PR TITLE
HTTP Path Filtering partial

### PR DESCRIPTION
**Pull request by aviramha**
Partial work for #1512 - would love to merge it already to avoid huge PR :)

### CODE REVIEW

File: mirrord/agent/src/steal/http/filter.rs [hunk 1]
```diff
@@ -35,6 +37,108 @@ const H2_PREFACE: &[u8; 14] = b"PRI * HTTP/2.0";
...
+    pub fn new_header_filter(header: Regex) -> Self {
+        Self::Header(HttpHeaderFilter(header))
+    }
+    #[allow(dead_code)]
+    pub fn new_path_filter(path: Regex) -> Self {
+        Self::Path(HttpPathFilter(path))
+    }
+}
+
+/// Struct to hold cache/hold/reuse the attributes that a [`Request`] can be matched upon
```
Comment (meowjesty) [inline mirrord/agent/src/steal/http/filter.rs:66]:
```suggestion
/// Struct to cache/hold/reuse the attributes that a [`Request`] can be matched upon
```

One "hold" too much?

File: mirrord/agent/src/steal/http/filter.rs [hunk 2]
```diff
@@ -35,6 +37,108 @@ const H2_PREFACE: &[u8; 14] = b"PRI * HTTP/2.0";
...
+                    error!(
+                        "Error while matching header: {header:?}, {:?}, {err:?}",
+                        self.0
+                    );
+                    return false;
+                }
+            }
+        }
+        false
+    }
```
Comment (meowjesty) [inline mirrord/agent/src/steal/http/filter.rs:129]:
```suggestion
request.headers().find_map(|header| self.0.is_match(header).unwrap_or_default()).unwrap_or_default()
```

Something along those lines? (sorry, I can't help myself not suggesting iterators)

File: mirrord/agent/src/steal/http/filter.rs [hunk 3]
```diff
@@ -35,6 +37,108 @@ const H2_PREFACE: &[u8; 14] = b"PRI * HTTP/2.0";
...
+        Self::Path(HttpPathFilter(path))
+    }
+}
+
+/// Struct to hold cache/hold/reuse the attributes that a [`Request`] can be matched upon
+pub(crate) struct RequestMatchCandidate<'a> {
+    /// The original request
+    request: &'a Request<Incoming>,
+    /// Headers in `k: v` format.
+    headers: OnceCell<Vec<String>>,
```
Comment (meowjesty) [inline mirrord/agent/src/steal/http/filter.rs:71]: Not understanding why we're using `OnceCell` here?

Can't we just initialize an empty `Vec`?

File: mirrord/agent/src/steal/http/hyper_handler.rs [hunk 4]
```diff
@@ -76,31 +79,19 @@ pub(super) struct RawHyperConnection {
...
-                        header
-                            .as_ref()
-                            .expect("The header value has to be convertible to `String`!"),
-                    )
-                    .expect("Something went wrong in the regex matcher!")
-                    .then_some(*filter.key())
-            })
-        })
+    let mut request = request.into();
+    filters.iter().find_map(|entry| {
```
Comment (meowjesty) [inline mirrord/agent/src/steal/http/hyper_handler.rs:88]: sweet, sweet iterators

**Review by meowjesty — CHANGES_REQUESTED** [review]
I assume the config for this is coming in another PR?

My first time seeing the `protocol_break` macro.

Only big thing that I'm not getting is the `OnceCell<Vec>>`, otherwise looks :+1:

File: mirrord/agent/src/steal/http/filter.rs (hunk 3 above)
Comment (aviramha) [reply mirrord/agent/src/steal/http/filter.rs:71]: The idea is to fill this lazily, only when needed. If the only filters that exist are path based, we wouldn't need to even initialize it.

File: mirrord/agent/src/steal/http/filter.rs (hunk 2 above)
Comment (aviramha) [reply mirrord/agent/src/steal/http/filter.rs:129]: I think with error handling the loop looks better

File: mirrord/agent/src/steal/http/filter.rs (hunk 3 above)
Comment (meowjesty) [reply mirrord/agent/src/steal/http/filter.rs:71]: hmm I get it. I'm not a fan of lazy initialization, but think this is too minor to require arguing against.

I can see that adds more meaning, just looks a bit weird (at least initially).

**Review by meowjesty — APPROVED** [review]
🐻 👍

**Pull request merged**